arangod/IResearch/IResearchDocument.cpp

```diff
@@ -693,7 +697,15 @@ bool DocumentPrimaryKey::read(irs::bytes_ref const& in) noexcept {
     return false;
   }
 
-  std::memcpy(_keys, in.c_str(), sizeof(_keys));
+  static_assert(
+    sizeof(TRI_voc_cid_t) == sizeof(TRI_voc_rid_t),
+    "sizeof(TRI_voc_cid_t) != sizeof(TRI_voc_rid_t)"
+  );
+
+  auto* begin = reinterpret_cast<TRI_voc_cid_t const*>(in.c_str());
```

Comment:
That's good catch in general, but I believe we good there:
`in` is a pointer to internal iresearch columstore buffer which is `std::basic_string<uint8_t>`, default allocator guarantees data will be aligned by `alignof(std::max_align_t)` which is greater or equal to `alignof(uint64_t)`.
The reason of why I changed the code is speed, current version works twice as faster and is used very frequently. So I believe we just have to leave a big red comment there. WDYT?

Reply:
I guess std::basic_string<T> is free to make its internal buffer start at any alignment it likes. Or is it guaranteed that `c_str()` returns an aligned pointer?
However I am sure this will work in reality right now, so a comment is sufficient IMHO.

Reply:
alignment of `std::string::c_str()` is guaranteed by `std::allocator`, the tricky part there is SBO. But since all known implementations has both `capacity` and `size` of type `size_t ` as `std::basic_string` members, alignment of `c_str()` should be equal to `alignof(size_t)` which is sufficient in our case of well

Reply:
Yes, I was thinking about SSO...
Clang's internal std::string implementation is quite different to the one found in g++. It does quite some smart stuff to have 22 or 23 chars for the SSO case I believe.